presto-thrift/src/main/java/com/facebook/presto/thrift/clientproviders/PrestoThriftClientProvider.java

```diff
@@ -0,0 +1,81 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.thrift.clientproviders;
+
+import com.facebook.nifty.client.FramedClientConnector;
+import com.facebook.presto.spi.HostAddress;
+import com.facebook.presto.spi.PrestoException;
+import com.facebook.presto.thrift.interfaces.client.ThriftPrestoClient;
+import com.facebook.presto.thrift.location.HostLocationProvider;
+import com.facebook.swift.service.ThriftClient;
+import com.google.common.net.HostAndPort;
+import io.airlift.units.Duration;
+
+import javax.inject.Inject;
+
+import java.util.List;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.TimeoutException;
+
+import static com.facebook.presto.thrift.ThriftErrorCode.CONNECTION_ERROR;
+import static com.google.common.base.Throwables.throwIfInstanceOf;
+import static java.util.Objects.requireNonNull;
+
+public class PrestoThriftClientProvider
+        implements PrestoClientProvider
```

Comment:
Also, regarding one of the previous comments about `connectToAnyHost` naming, I think the names you suggested closer represents what they do, so I renamed those as well.
Current names are: `ThriftServiceClient` (`anyHostClient()`, `selectedHostClient()`), `ThriftServiceClientProvider`, `RetryingThriftServiceClientProvider`.